OpenRA.Game/Support/BooleanExpression.cs

```diff
@@ -0,0 +1,245 @@
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace OpenRA.Support
+{
+	public class BooleanExpression
+	{
+		readonly HashSet<string> variables = new HashSet<string>();
+		public IEnumerable<string> Variables { get { return variables; } }
+
+		readonly Token[] tokens;
+
+		class Token
+		{
+			public readonly int Index;
+			public Token(int index)
+			{
+				Index = index;
+			}
+		}
+
+		class BinOpToken : Token { public BinOpToken(int index) : base(index) { } }
+		class AndToken : BinOpToken { public AndToken(int index) : base(index) { } }
+		class OrToken : BinOpToken { public OrToken(int index) : base(index) { } }
+		class XorToken : BinOpToken { public XorToken(int index) : base(index) { } }
+		class NotToken : Token { public NotToken(int index) : base(index) { } }
+		class OpenParenToken : Token { public OpenParenToken(int index) : base(index) { } }
+		class CloseParenToken : Token { public CloseParenToken(int index) : base(index) { } }
+		class VariableToken : Token
+		{
+			public readonly string Name;
+
+			public VariableToken(int index, string name)
+				: base(index)
+			{
+				Name = name;
+			}
+		}
+
+		public BooleanExpression(string expression)
+		{
+			var openParens = 0;
+			var closeParens = 0;
+			var tokens = new List<Token>();
+			for (var i = 0; i < expression.Length; i++)
+			{
+				switch (expression[i])
+				{
+					case '(':
+					{
+						tokens.Add(new OpenParenToken(i));
+						openParens++;
+						break;
+					}
+
+					case ')':
+					{
+						tokens.Add(new CloseParenToken(i));
+						if (++closeParens > openParens)
+							throw new InvalidDataException("Unmatched closing parenthesis at index {0}".F(i));
+
+						break;
+					}
+
+					case '!':
+					{
+						tokens.Add(new NotToken(i));
+						break;
+					}
+
+					case '^':
+					{
+						tokens.Add(new XorToken(i));
+						break;
+					}
+
+					default:
+					{
+						// Ignore whitespace
+						if (char.IsWhiteSpace(expression[i]))
+							break;
+
+						var token = ParseExtendedToken(expression, ref i);
+						tokens.Add(token);
+
+						var variable = token as VariableToken;
+						if (variable != null)
+							variables.Add(variable.Name);
+
+						break;
+					}
+				}
+			}
+
+			// Sanity check parsed tree
+			if (!tokens.Any())
+				throw new InvalidDataException("Empty expression");
+
+			if (closeParens != openParens)
+				throw new InvalidDataException("Mismatched opening and closing parenthesis");
+
+			for (var i = 0; i < tokens.Count - 1; i++)
+			{
+				if (tokens[i] is NotToken && !(tokens[i + 1] is VariableToken || tokens[i + 1] is OpenParenToken))
+					throw new InvalidDataException("Unexpected character `!` at index {0}".F(tokens[i].Index));
```

Comment:
Is there a concrete list of expected token types at the point of these exceptions?
If so those would be nice to include.